cilium unstable, crashing every few minutes/hours

### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Every few minutes/hours, Cilium will crash and restart. It works briefly (connectivity test passes, no visible issues with cluster networking) and then crashes again.

### Cilium Version

```
cilium-cli: 0.12.11 compiled with go1.19.3 on darwin/arm64
cilium image (default): v1.12.2
cilium image (stable): v1.12.5
cilium image (running): v1.13.0-rc2
```

### Kernel Version

Linux adenine 6.0.10 #1-NixOS SMP PREEMPT_DYNAMIC Sat Nov 26 08:27:56 UTC 2022 x86_64 GNU/Linux

### Kubernetes Version

Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.3", GitCommit:"434bfd82814af038ad94d62ebe59b133fcb50506", GitTreeState:"clean", BuildDate:"2022-10-12T10:49:09Z", GoVersion:"go1.19.2", Compiler:"gc", Platform:"linux/amd64"}

### Sysdump

File size too large for GitHub, uploaded to https://ocf.berkeley.edu/~njha/tmp/cilium-sysdump-20230105-114355.zip

### Relevant log output

```shell
panic: addr is not a valid IP address

goroutine 1688 [running]:
github.com/cilium/cilium/pkg/ip.MustAddrFromIP(...)
	/go/src/github.com/cilium/cilium/pkg/ip/ip.go:971
github.com/cilium/cilium/pkg/fqdn.(*DNSCache).LookupIP(0xc000db3920, {0x0?, 0x50?, 0x8?})
	/go/src/github.com/cilium/cilium/pkg/fqdn/cache.go:450 +0x154
github.com/cilium/cilium/daemon/cmd.(*Daemon).GetNamesOf(0x0?, 0x1b2d2de?, {0x0, 0x0, 0x0})
	/go/src/github.com/cilium/cilium/daemon/cmd/hubble.go:343 +0x69
github.com/cilium/cilium/pkg/hubble/parser/seven.(*Parser).Decode(0xc0016b41e0, 0xc0014236b0, 0xc0033ac480)
	/go/src/github.com/cilium/cilium/pkg/hubble/parser/seven/parser.go:98 +0x310
github.com/cilium/cilium/pkg/hubble/parser.(*Parser).Decode(0xc001252600, 0xc0022c9580)
	/go/src/github.com/cilium/cilium/pkg/hubble/parser/parser.go:128 +0x2fd
github.com/cilium/cilium/pkg/hubble/observer.(*LocalObserverServer).Start(0xc004240a20)
	/go/src/github.com/cilium/cilium/pkg/hubble/observer/local_observer.go:128 +0xf4
created by github.com/cilium/cilium/daemon/cmd.(*Daemon).launchHubble
	/go/src/github.com/cilium/cilium/daemon/cmd/hubble.go:173 +0x129c
```


### Anything else?

I suspect something like a workload in the cluster poisoning the cache or something, because at first this was only happening on one node, and now it has slowly spread to three nodes.

Looks like it has something to do with Hubble-- I'll try turning off Hubble and seeing if that works for now.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/daemon/cmd/hubble.go
+++ b/daemon/cmd/hubble.go
@@ -39,6 +39,7 @@ import (
 	"github.com/cilium/cilium/pkg/hubble/server"
 	"github.com/cilium/cilium/pkg/hubble/server/serveroption"
 	"github.com/cilium/cilium/pkg/identity"
+	ippkg "github.com/cilium/cilium/pkg/ip"
 	"github.com/cilium/cilium/pkg/ipcache"
 	"github.com/cilium/cilium/pkg/loadbalancer"
 	"github.com/cilium/cilium/pkg/logging"
@@ -340,7 +341,12 @@ func (d *Daemon) GetNamesOf(sourceEpID uint32, ip net.IP) []string {
 	if ep == nil {
 		return nil
 	}
-	names := ep.DNSHistory.LookupIP(ip)
+
+	addr, ok := ippkg.AddrFromIP(ip)
+	if !ok {
+		return nil
+	}
+	names := ep.DNSHistory.LookupIP(addr)
 
 	for i := range names {
 		names[i] = strings.TrimSuffix(names[i], ".")

--- a/pkg/fqdn/cache.go
+++ b/pkg/fqdn/cache.go
@@ -443,11 +443,11 @@ func (c *DNSCache) lookupByRegexpByTime(now time.Time, re *regexp.Regexp) (match
 // maintains the latest-expiring entry per-name per-IP. This means that multiple
 // names referrring to the same IP will expire from the cache at different times,
 // and only 1 entry for each name-IP pair is internally retained.
-func (c *DNSCache) LookupIP(ip net.IP) (names []string) {
+func (c *DNSCache) LookupIP(ip netip.Addr) (names []string) {
 	c.RLock()
 	defer c.RUnlock()
 
-	return c.lookupIPByTime(c.lastCleanup, ippkg.MustAddrFromIP(ip))
+	return c.lookupIPByTime(c.lastCleanup, ip)
 }
 
 // lookupIPByTime takes a timestamp for expiration comparisons, and is

--- a/pkg/fqdn/cache_test.go
+++ b/pkg/fqdn/cache_test.go
@@ -560,7 +560,7 @@ func (ds *DNSCacheTestSuite) TestOverlimitEntriesWithValidLimit(c *C) {
 	c.Assert(affectedNames, checker.DeepEquals, []string{"test.com"})
 
 	c.Assert(cache.Lookup("test.com"), HasLen, limit)
-	c.Assert(cache.LookupIP(net.ParseIP("1.1.1.1")), checker.DeepEquals, []string{"foo.bar"})
+	c.Assert(cache.LookupIP(netip.MustParseAddr("1.1.1.1")), checker.DeepEquals, []string{"foo.bar"})
 	c.Assert(cache.forward["test.com"][netip.MustParseAddr("1.1.1.1")], IsNil)
 	c.Assert(cache.Lookup("foo.bar"), HasLen, 1)
 	c.Assert(cache.Lookup("bar.foo"), HasLen, 1)

